### PR TITLE
Exclude DOM events from default handlers

### LINKED ISSUE #4600: There should be some way exclude DOM events fired by certain UIElements from default handlers
## Problem

ATM the widget takes over the DOM events like [mousedown](https://github.com/ckeditor/ckeditor5-widget/blob/master/src/widget.js#L112-L153) (also keydown, delete) and processes them as it thinks fit.

I realized when implementing https://github.com/ckeditor/ckeditor5/issues/1509 that this behaviour causes some issues. For instance, because `mousedown` is handled by the widget, I wasn't able to create any dynamic form elements (`<select>`) using React inside a `UIElement` in the widget. They render, that's alright, but you can't click them, focus etc. because widget wants to handle everything.

## What I thought was a solution

Unfortunately the problem is not as simple as "remove custom [event] handling when even target is `UIElement` or a descendant" because by doing this, we'd prevent widget selection when an `UIElement` is clicked, for instance a media widget uses `UIElement` to display the YouTube preview, and we definitely want clicking it to focus the widget.

I also tried to attach a native DOM `mousedown` listener on the `<select>` I rendered, and stop propagation of the event before it reaches the editable, before CKEditor's widget logick kicks in. I would make sense except that... React uses a single global listener for each event and it heavily depends on the fact that the event does bubble towards the root of the application. Stopping it like that simply breaks React so it's a dead–end.

## An actual solution

What I guess we need is some API to tell the widget that a particual `HTMLElement` should be excluded from custom event handling. 

How do we do that?

cc @jodator @Reinmar @mlewand

### PR DESCRIPTION
**Pull request by psmyrek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (engine): Allow excluding DOM events from default handlers. Closes #4600.

---

### Additional information

Event will be ignored only if its target element or any of its ancestors has the `data-cke-ignore-events` attribute.

### CODE REVIEW

**Comment by pomek** [comment]
Changes from the PR are required for https://github.com/ckeditor/ckeditor5/pull/8290.

**Review by mlewand — COMMENTED** [review]
Looks fine for the most part.

During the review I also checked some other observers. `ImageLoadObserver` - it fires for images nested in `data-cke-ignore-events="true"` elements. To be consistent I believe that the event should not be propagated as well. 

There are also following events:

* `DeleteObserver`
* `EnterObserver`
* `MutationObserver` (I have doubts whether we should ignore this one)

That to my surprise ignored in the widget (manual test), but are propagated from a `data-cke-ignore-events="true"` element that is not a widget. I created a simple `<section data-cke-ignore-events="true">` element ([pushed to a branch](https://github.com/ckeditor/ckeditor5/tree/i/4600-simple-element)). Editing such element is propagating the events, which make sense, because [there's no explicit check in them](https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-enter/src/enterobserver.js#L25-L42).

I also checked the `ClipboardObserver` - it works fine as it's inheriting from `DomEventObserver`.

Also the proposed changelog message should also name the attribute that you have introduced for easier discoverability.

----

To sum up:

* `DeleteObserver`
* `EnterObserver`

Needs to have the check for ignored attribute added.

I'm not sure about the following:

* `ImageLoadObserver` - I'm leaning towards KISS - ignoring the events fired within `data-cke-ignore-events`. Even though [it fires `layoutChanged` event](https://github.com/ckeditor/ckeditor5/blob/608baa9be5a1c8ae5600e8df9627c4f5b2cecef7/packages/ckeditor5-image/src/image/imageloadobserver.js#L47) (by using `data-cke-ignore-events` implementer should take the responsibility).
* `MutationObserver` - IDK about implications of changing that.

@reinmar @jodator any thoughts on two above? 👆

**Comment by Reinmar** [comment]
As for \`EnterObserver\` and \`DeleteObserver\` – isn't `keydown` (on which they are based) filtered already? That event is our event and it may be already checked in `KeyObserver`. Write tests for both observers and you'll know :)

> *   `MutationObserver` (I have doubts whether we should ignore this one)

Yes and no. Yes, because it would be good to have support for `data-cke-ignore-events` consistent. However, it's not necessary today as in all cases I know and all I'd consider correct, `data-cke-ignore-events` is used inside a `UIElement` or `RawElement` and `MutationObserver` ignore mutations coming from these elements. So, we could add a `data-cke-ignore-events` check in it, but I think that that's not really necessary today.

**Comment by mlewand** [comment]
We'll merge the changes now and fix the review notes in a followup task as Raw HTML embedding is depending on this feature.

**Comment by Reinmar** [comment]
BTW, I checked the manual test out of curiosity and I think it's a bit complicated. I could not quickly verify whether it works the way I assumed as there's quite a lot to understand. For example:

> Clicking on a button dispatches the `click` event (the `Dispatched click event` message should be logged), but `Received click event` shouldn't be present in console.

Why? Why is click dispatched logged? Why should I ignore it?

I think that a manual test should be relatively easy to understand and verify the results without going deep into how it's functioning. 

Another thing is that we may need to fix the blur/focus thing – as @oleq pointed out during discussions around HTML embedding, the editor should not lose focus while the user is inside an ignored element. Could you open a ticket for this?

**Comment by mlewand** [comment]
The manual test indeed needs some simplification e.g. see https://github.com/ckeditor/ckeditor5/pull/8243#discussion_r508216053, https://github.com/ckeditor/ckeditor5/pull/8243#discussion_r508220498.

File: packages/ckeditor5-engine/tests/view/observer/domeventobserver.js [hunk 1]
```diff
@@ -122,6 +122,28 @@ describe( 'DomEventObserver', () => {
 		expect( evtSpy.called ).to.be.false;
 	} );
 
+	it( 'should not fire event if target is ignored', () => {
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/view/observer/domeventobserver.js:125]: Done in https://github.com/ckeditor/ckeditor5/pull/8305.

File: packages/ckeditor5-engine/tests/view/observer/domeventobserver.js [hunk 2]
```diff
@@ -122,6 +122,28 @@ describe( 'DomEventObserver', () => {
...
+		view.attachDomRoot( domElement );
+		view.addObserver( ClickObserver );
+		viewDocument.on( 'click', evtSpy );
+
+		domElement.dispatchEvent( domEvent );
+
+		expect( ignoreEventFromTargetStub.called ).to.be.true;
+		expect( evtSpy.called ).to.be.false;
+
+		ignoreEventFromTargetStub.restore();
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/view/observer/domeventobserver.js:144]: Done in https://github.com/ckeditor/ckeditor5/pull/8305.

File: packages/ckeditor5-engine/tests/manual/tickets/4600/1.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	view.domConverter
+		.viewToDom( container )
+		.querySelectorAll( 'button' )
+		.forEach( button => {
+			button.addEventListener( 'click', event => {
+				if ( !event.isTrusted ) {
+					return;
+				}
+
+				console.log( `Dispatched ${ eventName } event.` );
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/manual/tickets/4600/1.js:151]: Done in https://github.com/ckeditor/ckeditor5/pull/8305.

File: packages/ckeditor5-engine/tests/manual/tickets/4600/1.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+			events.forEach( eventName => {
+				this.listenTo( view.document, eventName, () => {
+					console.log( `Received ${ eventName } event.` );
+				} );
+			} );
+		} );
+	}
+}
+
+class SimpleWidgetUI extends Plugin {}
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/manual/tickets/4600/1.js:116]: Done in https://github.com/ckeditor/ckeditor5/pull/8305.

File: packages/ckeditor5-engine/tests/manual/tickets/4600/1.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+		.viewToDom( container )
+		.querySelectorAll( 'button' )
+		.forEach( button => {
+			button.addEventListener( 'click', event => {
+				if ( !event.isTrusted ) {
+					return;
+				}
+
+				console.log( `Dispatched ${ eventName } event.` );
+				button.dispatchEvent( new Event( eventName, { bubbles: true } ) );
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/manual/tickets/4600/1.js:152]: The `domTarget` for image element is `undefined` so I'm showing there `eventData.domTarget || eventData.target`.

Done in https://github.com/ckeditor/ckeditor5/pull/8305.

File: packages/ckeditor5-engine/tests/manual/tickets/4600/1.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+class SimpleWidget extends Plugin {
+	static get requires() {
+		return [ SimpleWidgetEditing, SimpleWidgetUI ];
+	}
+}
+
+ClassicEditor
+	.create( document.querySelector( '#editor' ), {
+		plugins: [ Essentials, SimpleWidget ]
```
[the remark this answers is not in the document]
Comment (psmyrek) [reply packages/ckeditor5-engine/tests/manual/tickets/4600/1.js:126]: I've added the `Paragraph` plugin (rather than `Plugin` plugin) into `plugins` array.

Done in https://github.com/ckeditor/ckeditor5/pull/8305.

**Comment by psmyrek** [comment]
I've changed the manual test description and information that is logged in the console in https://github.com/ckeditor/ckeditor5/pull/8305. I hope now it's simpler and more clear what is going on there.

File: packages/ckeditor5-engine/tests/manual/tickets/4600/1.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mlewand) [reply packages/ckeditor5-engine/tests/manual/tickets/4600/1.js:126]: Silly mistake from my side, naturally I meant imported Paragraph plugin there 👍
